File: packages/zbb/lib/monorepo/Workspace.ts
/**
 * Workspace discovery, dependency graph, and topological sort for npm workspaces monorepos.
 */

import { readFileSync, existsSync } from 'node:fs';
import { join, resolve, relative } from 'node:path';
import { globSync } from 'node:fs';

// ── Types ────────────────────────────────────────────────────────────

export interface WorkspacePackage {
  /** npm package name, e.g. "@zerobias-com/hub-core" */
  name: string;
  /** Absolute path to the package directory */
  dir: string;
  /** Relative path from repo root, e.g. "core" */
  relDir: string;
  /** Version from package.json */
  version: string;
  /** Whether the package is private */
  private: boolean;
  /** npm scripts from package.json */
  scripts: Record<string, string>;
  /** Names of workspace packages this depends on */
  internalDeps: string[];
  /** Raw package.json content */
  packageJson: Record<string, any>;
}

export interface DependencyGraph {
  /** All workspace packages, keyed by npm name */
  packages: Map<string, WorkspacePackage>;
  /** Reverse adjacency: package name → set of package names that depend on it */
  dependents: Map<string, Set<string>>;
  /** Topological sort order (leaves/dependencies first, dependents last) */
  buildOrder: string[];
}

// ── Discovery ────────────────────────────────────────────────────────

/**
 * Discover all workspace packages from the root package.json.
 */
export function discoverWorkspaces(repoRoot: string): Map<string, WorkspacePackage> {
  const rootPkgPath = join(repoRoot, 'package.json');
  if (!existsSync(rootPkgPath)) {
    throw new Error(`No package.json found at ${repoRoot}`);
  }

  const rootPkg = JSON.parse(readFileSync(rootPkgPath, 'utf-8'));
  const workspaceGlobs: string[] = rootPkg.workspaces ?? [];
  if (workspaceGlobs.length === 0) {
    throw new Error('No workspaces defined in root package.json');
  }

  // Resolve workspace globs to directories
  const packageDirs: string[] = [];
  for (const glob of workspaceGlobs) {
    if (glob.includes('*')) {
      // Expand glob patterns like "packages/*"
      const expanded = globSync(glob, { cwd: repoRoot });
      for (const entry of expanded) {
        const absDir = resolve(repoRoot, entry);
        if (existsSync(join(absDir, 'package.json'))) {
          packageDirs.push(absDir);
        }
      }
    } else {
      // Plain directory name like "core" or "server"
      const absDir = resolve(repoRoot, glob);
      if (existsSync(join(absDir, 'package.json'))) {
        packageDirs.push(absDir);
      }
    }
  }

  // Read each workspace package.json
  const packages = new Map<string, WorkspacePackage>();
  const nameSet = new Set<string>();

  // First pass: collect all package names
  const pkgInfos: { dir: string; pkg: Record<string, any> }[] = [];
  for (const dir of packageDirs) {
    const pkgPath = join(dir, 'package.json');
    const pkg = JSON.parse(readFileSync(pkgPath, 'utf-8'));
    pkgInfos.push({ dir, pkg });
    nameSet.add(pkg.name);
  }

  // Second pass: resolve internal deps
  for (const { dir, pkg } of pkgInfos) {
    const allDeps = {
      ...pkg.dependencies,
      ...pkg.devDependencies,
    };

    const internalDeps: string[] = [];
    for (const depName of Object.keys(allDeps)) {
      if (nameSet.has(depName)) {
        internalDeps.push(depName);
      }
    }

    const wp: WorkspacePackage = {
      name: pkg.name,
      dir,
      relDir: relative(repoRoot, dir),
      version: pkg.version ?? '0.0.0',
      private: pkg.private ?? false,
      scripts: pkg.scripts ?? {},
      internalDeps,
      packageJson: pkg,
    };

    packages.set(pkg.name, wp);
  }

  return packages;
}

// ── Dependency Graph ─────────────────────────────────────────────────

/**
 * Build the dependency graph from discovered workspace packages.
 */
export function buildDependencyGraph(packages: Map<string, WorkspacePackage>): DependencyGraph {
  // Build reverse adjacency list (dependents)
  const dependents = new Map<string, Set<string>>();
  for (const [name] of packages) {
    dependents.set(name, new Set());
  }

  for (const [name, pkg] of packages) {
    for (const dep of pkg.internalDeps) {
      dependents.get(dep)?.add(name);
    }
  }

  // Topological sort using Kahn's algorithm
  const buildOrder = topologicalSort(packages, dependents);

  return { packages, dependents, buildOrder };
}

/**
 * Kahn's algorithm: BFS topological sort. Returns packages in dependency order
 * (leaves first, so they can be built before their dependents).
 */
function topologicalSort(
  packages: Map<string, WorkspacePackage>,
  dependents: Map<string, Set<string>>,
): string[] {
  // Compute in-degree (number of internal deps each package has)
  const inDegree = new Map<string, number>();
  for (const [name, pkg] of packages) {
    // Only count deps that are actually in the workspace
    const count = pkg.internalDeps.filter(d => packages.has(d)).length;
    inDegree.set(name, count);
  }

  // Start with packages that have no internal deps
  const queue: string[] = [];
  for (const [name, degree] of inDegree) {
    if (degree === 0) queue.push(name);
  }

  const sorted: string[] = [];

  while (queue.length > 0) {
    const current = queue.shift()!;
    sorted.push(current);

    // Reduce in-degree for all dependents of current
    const deps = dependents.get(current) ?? new Set();
    for (const dependent of deps) {
      const newDegree = (inDegree.get(dependent) ?? 0) - 1;
      inDegree.set(dependent, newDegree);
      if (newDegree === 0) {
        queue.push(dependent);
      }
    }
  }

  // Detect cycles
  if (sorted.length !== packages.size) {
    const remaining = [...packages.keys()].filter(n => !sorted.includes(n));
    throw new Error(
      `Circular dependency detected among workspace packages: ${remaining.join(', ')}`
    );
  }

  return sorted;
}

/**
 * Get all transitive dependents of a package (BFS through reverse adjacency).
 * Does NOT include the starting package itself.
 */
export function getTransitiveDependents(
  packageName: string,
  graph: DependencyGraph,
): Set<string> {
  const visited = new Set<string>();
  const queue = [packageName];

  while (queue.length > 0) {
    const current = queue.shift()!;
    const deps = graph.dependents.get(current) ?? new Set();
    for (const dep of deps) {
      if (!visited.has(dep)) {
        visited.add(dep);
        queue.push(dep);
      }
    }
  }

  return visited;
}

/**
 * Filter and sort a set of package names by the graph's build order.
 */
export function sortByBuildOrder(names: Set<string>, graph: DependencyGraph): string[] {
  return graph.buildOrder.filter(n => names.has(n));
}

File: packages/build-tools/src/main/kotlin/com/zerobias/buildtools/monorepo/Workspace.kt
package com.zerobias.buildtools.monorepo

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import java.io.File

/**
 * Workspace discovery, dependency graph, and topological sort for npm
 * workspace monorepos. Pure Kotlin — no Gradle dependency.
 *
 * Mirrors `org/util/packages/zbb/lib/monorepo/Workspace.ts`.
 *
 * Used by:
 * - `zb.monorepo-settings.gradle.kts` settings plugin (calls `discoverWorkspaces`
 *   and includes each as a Gradle subproject via `include(":path")`)
 * - `zb.monorepo-base.gradle.kts` (exposes the dep graph as a BuildService for
 *   ChangeDetector and the root aggregator tasks)
 */
data class WorkspacePackage(
    /** npm package name, e.g. "@zerobias-com/util-core" */
    val name: String,
    /** Absolute path to the package directory */
    val dir: File,
    /** Relative path from repo root, e.g. "packages/core" */
    val relDir: String,
    /** Version from package.json */
    val version: String,
    /** Whether the package is private */
    val private: Boolean,
    /** npm scripts from package.json */
    val scripts: Map<String, String>,
    /** Names of workspace packages this package depends on */
    val internalDeps: List<String>,
    /** Raw package.json content */
    val packageJson: Map<String, Any?>,
)

data class DependencyGraph(
    /** All workspace packages, keyed by npm package name */
    val packages: Map<String, WorkspacePackage>,
    /** Reverse adjacency: package name → set of names that depend on it */
    val dependents: Map<String, Set<String>>,
    /** Topological sort (leaves first, dependents last) */
    val buildOrder: List<String>,
)

object Workspace {
    private val mapper = ObjectMapper().registerKotlinModule()

    /**
     * Discover all workspace packages from the root package.json.
     *
     * Reads root `package.json`, expands the `workspaces` field (supports
     * glob patterns like `packages/&#42;` and literal paths), and parses each
     * workspace package's `package.json`.
     *
     * Throws if the root package.json is missing or has no workspaces.
     */
    fun discoverWorkspaces(repoRoot: File): Map<String, WorkspacePackage> {
        val rootPkgFile = File(repoRoot, "package.json")
        if (!rootPkgFile.exists()) {
            throw IllegalStateException("No package.json found at ${repoRoot.absolutePath}")
        }

        val rootPkg: Map<String, Any?> = mapper.readValue(rootPkgFile)
        @Suppress("UNCHECKED_CAST")
        val workspaceGlobs = (rootPkg["workspaces"] as? List<String>) ?: emptyList()
        if (workspaceGlobs.isEmpty()) {
            throw IllegalStateException("No workspaces defined in root package.json")
        }

        // Resolve workspace globs to package directories
        val packageDirs = mutableListOf<File>()
        for (glob in workspaceGlobs) {
            if (glob.contains("*")) {
                // Glob expansion: split on /*, walk
                val baseStr = glob.substringBefore("/*")
                val baseDir = File(repoRoot, baseStr)
                if (baseDir.isDirectory) {
                    baseDir.listFiles { f -> f.isDirectory && File(f, "package.json").exists() }
                        ?.forEach { packageDirs.add(it) }
                }
            } else {
                val absDir = File(repoRoot, glob)
                if (File(absDir, "package.json").exists()) {
                    packageDirs.add(absDir)
                }
            }
        }

        // First pass: read all package.json files, collect names
        data class PkgInfo(val dir: File, val pkg: Map<String, Any?>)
        val infos = packageDirs.mapNotNull { dir ->
            try {
                val pkgFile = File(dir, "package.json")
                val pkg: Map<String, Any?> = mapper.readValue(pkgFile)
                PkgInfo(dir, pkg)
            } catch (_: Exception) {
                null
            }
        }
        val nameSet = infos.mapNotNull { it.pkg["name"] as? String }.toSet()

        // Second pass: build WorkspacePackage objects with internal deps resolved
        val packages = linkedMapOf<String, WorkspacePackage>()
        for (info in infos) {
            val name = info.pkg["name"] as? String ?: continue
            @Suppress("UNCHECKED_CAST")
            val deps = (info.pkg["dependencies"] as? Map<String, Any?>) ?: emptyMap()
            @Suppress("UNCHECKED_CAST")
            val devDeps = (info.pkg["devDependencies"] as? Map<String, Any?>) ?: emptyMap()
            val allDepNames = (deps.keys + devDeps.keys).toSet()
            val internalDeps = allDepNames.filter { it in nameSet }

            @Suppress("UNCHECKED_CAST")
            val scriptsRaw = (info.pkg["scripts"] as? Map<String, Any?>) ?: emptyMap()
            val scripts = scriptsRaw.mapNotNull { (k, v) ->
                if (v is String) k to v else null
            }.toMap()

            packages[name] = WorkspacePackage(
                name = name,
                dir = info.dir,
                relDir = info.dir.relativeTo(repoRoot).path,
                version = info.pkg["version"] as? String ?: "0.0.0",
                private = info.pkg["private"] as? Boolean ?: false,
                scripts = scripts,
                internalDeps = internalDeps,
                packageJson = info.pkg,
            )
        }

        return packages
    }

    /**
     * Build the dependency graph from discovered workspace packages.
     */
    fun buildDependencyGraph(packages: Map<String, WorkspacePackage>): DependencyGraph {
        // Reverse adjacency
        val dependents = mutableMapOf<String, MutableSet<String>>()
        for (name in packages.keys) dependents[name] = mutableSetOf()
        for ((name, pkg) in packages) {
            for (dep in pkg.internalDeps) {
                dependents[dep]?.add(name)
            }
        }

        // Topological sort (Kahn's algorithm: leaves first)
        val buildOrder = topologicalSort(packages, dependents)

        return DependencyGraph(
            packages = packages,
            dependents = dependents.mapValues { it.value.toSet() },
            buildOrder = buildOrder,
        )
    }

    private fun topologicalSort(
        packages: Map<String, WorkspacePackage>,
        dependents: Map<String, Set<String>>,
    ): List<String> {
        // Compute in-degree (number of internal deps each package has,
        // restricted to packages in the workspace)
        val inDegree = mutableMapOf<String, Int>()
        for ((name, pkg) in packages) {
            inDegree[name] = pkg.internalDeps.count { it in packages }
        }

        // Start with all packages with no internal deps
        val queue = ArrayDeque<String>()
        for ((name, degree) in inDegree) {
            if (degree == 0) queue.add(name)
        }

        val sorted = mutableListOf<String>()
        while (queue.isNotEmpty()) {
            val current = queue.removeFirst()
            sorted.add(current)
            for (dependent in dependents[current] ?: emptySet()) {
                val newDegree = (inDegree[dependent] ?: 0) - 1
                inDegree[dependent] = newDegree
                if (newDegree == 0) queue.add(dependent)
            }
        }

        if (sorted.size != packages.size) {
            val remaining = packages.keys - sorted.toSet()
            throw IllegalStateException(
                "Circular dependency detected among workspace packages: ${remaining.joinToString(", ")}"
            )
        }

        return sorted
    }

    /**
     * Get all transitive dependents of a package via BFS through the reverse graph.
     * Does NOT include the starting package itself.
     */
    fun getTransitiveDependents(packageName: String, graph: DependencyGraph): Set<String> {
        val visited = mutableSetOf<String>()
        val queue = ArrayDeque<String>()
        queue.add(packageName)
        while (queue.isNotEmpty()) {
            val current = queue.removeFirst()
            for (dep in graph.dependents[current] ?: emptySet()) {
                if (dep !in visited) {
                    visited.add(dep)
                    queue.add(dep)
                }
            }
        }
        return visited
    }

    /**
     * Filter and sort a set of package names by the graph's build order.
     */
    fun sortByBuildOrder(names: Set<String>, graph: DependencyGraph): List<String> {
        return graph.buildOrder.filter { it in names }
    }
}
